<a href="https://colab.research.google.com/github/lyntos/Python-MS/blob/main/06_Coding_Exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

11.1 What does the following code do?
```
In [1]: def mystery(a, b):
...: if b == 1:
...: return a
...: else:
...: return a + mystery(a, b - 1)
...:
In [2]: mystery(2, 10)
Out[2]: ?????
```

Answer: The code above is a recursive function. It will repeatedly add `a` to itself "b times".

So `mystery(2,10)` will keep adding a to itself for 10 times until it meet the condition of b being equal to 1 - `mystery(2, 1)`. This will result in an output of `20`

In [ ]:
def mystery(a, b):
  if b == 1:
    return a
  else:
    print(a + mystery(a, b - 1))
    return a + mystery(a, b - 1)

mystery(2, 10)

11.2 Find the logic error(s) in the following recursive function, and
explain how to correct it (them). This function should find the sum of the
values from 0 to n.
```
In [3]: def sum(n):
...: if n == 0:
...: return 0
...: else:
...: return n + sum(n)
...:
```

Answer: The function `sum(n)` fails one of the basic rules of recursion - the recursive call has to change or reduce the value of the previous argument passed into the function call again.

`return n + sum(n)` does not do this, so the function will end in an infinite error (RecursionError).
The block below shows an updated code

In [ ]:
"""Fixed code"""
def sum(n):
  if n == 0:
    return 0
  else:
    return n + sum(n - 1)

11.3 What does the following code do?
```
In [4]: def mystery(a_array, size):
...: if size == 1:
...: return a_array[0]
...: else:
...: return a_array[size - 1] +
mystery(a_array, size - 1)
...:
In [5]: import numpy as np
In [6]: numbers = np.arange(1, 11)
In [7]: mystery(numbers, len(numbers))
Out[7]: ?????
```

Answer: It is a recursive function that:
1. Takes an array (a_array)
2. Takes its size (size)
3. Recursively adds the last element to the sum of the remaining elements.

For the given example of `mystery(numbers, len(numbers))` where numbers is a range from 1 to 10, the code runs `mystery(numbers, 10)` resulting in
```
10 + mystery(numbers, 9)
9 + mystery(numbers, 8)
...
1
```
which calculates
```
1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 + 10
```
resulting in `55` as the output.

11.4 In Section 11.3, we presented a recursive factorial function.
What happens if you remove the if statement from the factorial
function, then call the function?

Answer: The code will throw an error - specifically an `IndexError` since it will attempt to subtract 1 from the size (1) and attempt to fetch an index (-1) which does not exist.

In [ ]:
def mystery(a_array, size):
  if size == 1:
    return a_array[0]
  else:
    return a_array[size - 1] + mystery(a_array, size - 1)

import numpy as np
numbers = np.arange(1, 11)
mystery(numbers, len(numbers))

IndexError: index -11 is out of bounds for axis 0 with size 10

11.5 (Recursive power Function) Write a recursive function
power(base, exponent) that, when called, returns
```
base ^exponent
```
For example, power(3,4) = 3 * 3 * 3 * 3. Assume that exponent is
an integer greater than or equal to 1. Hint: The recursion step should use
the relationship
```
base^e = base · base^e-1
```
and the terminating condition occurs when exponent is equal to 1,
because
```
base^1 = base
```
Incorporate this function into a program that enables the user to enter the
base and exponent.

In [ ]:
def power(base, exponent):
  """
  Description: Takes a base number and an exponent and returns the base
  to the power of the exponent.
  """
  if exponent == 1:
    return base
  else:
    return base * power(base, exponent - 1)

print(power(3, 4))

11.6 (Recursive Fibonacci Modification) Modify Section 11.4’s recursive
fibonacci function to keep track of the total number of function calls.
Display the number of calls for fibonacci(10), fibonacci(20)
and fibonacci(30).

In [ ]:
call_count = 0

def fibonacci(n):
  global call_count
  call_count += 1

  if n <= 1:
    return n
  else:
    return fibonacci(n - 1) + fibonacci(n - 2)


# Test values
for n in [10, 20, 30]:
  call_count = 0
  result = fibonacci(n)
  print(f"fibonacci({n}) = {result}")
  print(f"Total function calls: {call_count}\n")

fibonacci(10) = 55
Total function calls: 177

fibonacci(20) = 6765
Total function calls: 21891

fibonacci(30) = 832040
Total function calls: 2692537



11.7 (Improving Recursive Fibonacci Performance: Memoization)
Research the performance enhancement technique called memoization.
Modify Section 11.4’s recursive fibonacci function to incorporate
memoization. Compare the performance of both versions for
`fibonacci(10)`, `fibonacci(20)` and `fibonacci(30)`.

In [ ]:
call_count = 0

def fibonacci_memo(n, memo={}):
  global call_count
  call_count += 1

  if n in memo:
    return memo[n]

  if n <= 1:
    memo[n] = n
  else:
    memo[n] = fibonacci_memo(n - 1, memo) + fibonacci_memo(n - 2, memo)

  return memo[n]


# Compare versions
for n in [10, 20, 30]:
  call_count = 0
  result = fibonacci_memo(n, {})
  print(f"Memoized fibonacci({n}) = {result}")
  print(f"Total function calls: {call_count}\n")

Memoized fibonacci(10) = 55
Total function calls: 19

Memoized fibonacci(20) = 6765
Total function calls: 39

Memoized fibonacci(30) = 832040
Total function calls: 59



11.8 (Visualizing Recursion) It’s interesting to watch recursion “in action.”
Modify the recursive factorial function presented in this chapter to
print its local variable and recursive-call parameter. For each recursive
call, display the outputs on a separate line and add a level of indentation.
Make the outputs clear, interesting and meaningful. Your goal here is to
design and implement an output format that makes it easier to understand
recursion.

In [ ]:
def factorial(n, depth=0):
  indent = "  " * depth

  print(f"{indent}Entering factorial({n})")

  if n == 1:
    print(f"{indent}Base case reached: factorial(1) = 1")
    return 1
  else:
    print(f"{indent}Computing {n} * factorial({n-1})")

    result = n * factorial(n - 1, depth + 1)

    print(f"{indent}Returning {result} from factorial({n})")
    return result


# Test
print("\nFinal Result:", factorial(5))

Entering factorial(5)
Computing 5 * factorial(4)
  Entering factorial(4)
  Computing 4 * factorial(3)
    Entering factorial(3)
    Computing 3 * factorial(2)
      Entering factorial(2)
      Computing 2 * factorial(1)
        Entering factorial(1)
        Base case reached: factorial(1) = 1
      Returning 2 from factorial(2)
    Returning 6 from factorial(3)
  Returning 24 from factorial(4)
Returning 120 from factorial(5)

Final Result: 120


11.9 (Greatest Common Divisor) The greatest common divisor of integers
x and y is the largest integer that evenly divides into both x and y. Write
and test a recursive function gcd that returns the greatest common divisor
of x and y. The gcd of x and y is defined recursively as follows: If y is
equal to 0, then gcd(x, y) is x; otherwise, gcd(x, y) is gcd(y, x %
y), where % is the remainder operator.

In [ ]:
"""
  Description: Returns the greatest common divisor of the given integers
"""
def gcd(x, y):
  if y == 0:
    return x
  else:
    return gcd(y, x % y)

print(gcd(48, 18))

6


11.10 (Palindromes) A palindrome is a string that is spelled the same way
forward and backward. Some examples of palindromes are “radar,” “able
was i ere i saw elba” and (if spaces are ignored) “a man a plan a canal
panama.” Write a recursive test_palindrome function that returns
True if the string stored in an array is a palindrome and False
otherwise. The function should ignore spaces and punctuation in the
string.

In [ ]:
import re

def test_palindrome(text):
  """
  Description: Returns True if the given string is a palindrome and False otherwise.
  """
  # Remove spaces, punctuation, make lowercase
  cleaned = re.sub(r'[^a-zA-Z0-9]', '', text).lower()

  def recursive_check(s):
    # Base case: length 0 or 1
    if len(s) <= 1:
      return True

    # If first and last characters differ = not palindrome
    if s[0] != s[-1]:
      return False

    # Recursive step: check inner substring
    return recursive_check(s[1:-1])

  return recursive_check(cleaned)

print(test_palindrome("radar"))
print(test_palindrome("Able was I ere I saw Elba"))
print(test_palindrome("A man, a plan, a canal, Panama!"))
print(test_palindrome("hello"))

True
True
True
False


11.11 (Eight Queens) A puzzler for chess buffs is the Eight Queens
problem, which asks: Is it possible to place eight queens on an empty
chessboard so that no queen is “attacking” any other (i.e., no two queens
are in the same row, in the same column or along the same diagonal)? For
example, if a queen is placed in the upper-left corner of the board, no other
queens could be placed in any of the marked squares shown in the
following figure. Solve the problem recursively. [Hint: Your solution
should begin with the first column and look for a location in that column
where a queen can be placed—initially, place the queen in the first row.
The solution should then recursively search the remaining columns. In the
first few columns, there will be several locations where a queen may be
placed. Take the first available location. If a column is reached with no
possible location for a queen, the program should return to the previous
column, and move the queen in that column to a new row. This continuous
backing up and trying new alternatives is an example of recursive
backtracking.]

11.12 (Towers of Hanoi) In this chapter, you studied functions that can be
easily implemented both recursively and iteratively. In this exercise, we
present a problem whose recursive solution demonstrates the elegance of
recursion, and whose iterative solution may not be as apparent.
The Towers of Hanoi is one of the most famous classic problems every
budding computer scientist must grapple with. Legend has it that in a
temple in the Far East, priests are attempting to move a stack of golden
disks from one diamond peg to another. The following diagram shows the
pegs with four disks on peg 1.

The initial stack has 64 disks threaded onto one peg and arranged from
bottom to top by decreasing size. The priests are attempting to move the
stack from one peg to another under the constraints that exactly one disk is
moved at a time and at no time may a larger disk be placed above a
smaller disk. Three pegs are provided, one being used for temporarily
holding disks. Supposedly, the world will end when the priests complete
their task, so there is little incentive for us to facilitate their efforts.
Let’s assume that the priests are attempting to move the disks from peg 1
to peg 3. We wish to develop an algorithm that prints the precise sequence
of peg-to-peg disk transfers.
If we were to approach this problem with conventional functions, we
would rapidly find ourselves hopelessly knotted up in managing the disks.
Instead, attacking this problem with recursion in mind allows the steps to
be simple. Moving n disks can be viewed in terms of moving only n – 1
disks (hence, the recursion), as follows:
1. Move n – 1 disks from peg 1 to peg 2, using peg 3 as a temporary
holding area.
2. Move the last disk (the largest) from peg 1 to peg 3.
3. Move the n – 1 disks from peg 2 to peg 3, using peg 1 as a
temporary holding area.
The process ends when the last task involves moving n = 1 disk (i.e., the
base case). This task is accomplished by simply moving the disk, without
the need for a temporary holding area. Write a program to solve the
Towers of Hanoi problem. Use a recursive function with four parameters:
1. The number of disks to be moved.
2. The peg on which these disks are initially threaded.
3. The peg to which this stack of disks is to be moved.
4. The peg to be used as a temporary holding area.
Display the precise instructions for moving the disks from the starting peg
to the destination peg. To move a stack of three disks from peg 1 to peg 3,
the program displays the following moves:
1 → 3 (This means move one disk from peg 1
to peg 3.)
1 → 2
3 → 2
1 → 3
2 → 1
2 → 3
1 → 3

11.13 (Logarithmic Portion of Big O) What key aspect of both the binary
search and the merge sort accounts for the logarithmic portion of their
respective Big Os?

Answer:
Both repeated half their respective input.

Halving the problem size each step is the logarithmic growth.

That divide-by-2 behavior is what creates the log(n) term.

11.14 (Comparing Insertion Sort to Merge Sort) In what sense is the
insertion sort superior to the merge sort? In what sense is the merge sort
superior to the insertion sort?

Answer:
Insertion sort is better for small or nearly sorted arrays.

Merge sort is better for large datasets and predictable performance.

11.15 (Merge Sort: Sorting Subarrays) In the text, we say that after the
merge sort splits the array into two subarrays, it then sorts these two
subarrays and merges them. Why might someone be puzzled by our
statement that “it then sorts these two subarrays”?

Answer:

Because merge sort does not directly sort the two halves immediately.

It:
	1.	Splits the array into two halves.
	2.	Recursively calls merge sort on each half.
	3.	Eventually reaches subarrays of size 1 (which are already sorted).
	4.	Then merges them back together in sorted order.

11.16 (Timing Sorting Algorithms) Remove the output statements from
the functions selection_sort, insertion_sort and
merge_sort defined in this chapter, then import each example’s sourcecode
file into IPython. Create a 100,000-element array of random integers
named data1 and make two additional copies of the array (data2 and
data3) by calling method copy on the original array. Next, use
%timeit as follows to compare the performance of each sorting
algorithm:
```
%timeit -n 1 -r 1
selectionsort.selection_sort(data1)
%timeit -n 1 -r 1
insertionsort.insertion_sort(data2)
%timeit -n 1 -r 1
mergesort.merge_sort(data3)
```
Do the selection_sort and insertion_sort take approximately
the same amount of time? Is merge_sort much faster?

11.17 (Bucket Sort) A bucket sort begins with a one-dimensional array of
positive integers to be sorted and a two-dimensional array of integers with
rows indexed from 0 to 9 and columns indexed from 0 to n – 1, where n is
the number of values to be sorted. Each row of the two-dimensional array
is referred to as a bucket. Write a class named BucketSort containing a
function called sort that operates as follows:
1. Place each value of the one-dimensional array into a row of the
bucket array, based on the value’s “ones” (rightmost) digit. For
example, 97 is placed in row 7, 3 is placed in row 3, and 100 is
placed in row 0. This procedure is called a distribution pass.
2. Loop through the bucket array row by row, and copy the values
back to the original array. This procedure is called a gathering
pass. The new order of the preceding values in the onedimensional
array is 100, 3 and 97.
3. Repeat this process for each subsequent digit position (tens,
hundreds, thousands, etc.). On the second (tens digit) pass, 100 is
placed in row 0, 3 is placed in row 0 (because 3 has no tens
digit), and 97 is placed in row 9. After the gathering pass, the
order of the values in the one-dimensional array is 100, 3 and 97.
On the third (hundreds digit) pass, 100 is placed in row 1, 3 is
placed in row 0, and 97 is placed in row 0 (after the 3). After this
last gathering pass, the original array is in sorted order.
The two-dimensional array of buckets is 10 times the length of
the integer array being sorted. This sorting technique provides
better performance than a selection and insertion sorts but
requires much more memory—the selection and insertion sorts
require space for only one additional element of data. This
comparison is an example of a space/time trade-off: The bucket
sort uses more memory than the selection and insertion sorts, but
performs better. This version of the bucket sort requires copying
all the data back to the original array on each pass. Another
possibility is to create a second two-dimensional bucket array
and repeatedly swap the data between the two bucket arrays.

11.18 (Recursive Binary Search) Modify this chapter’s binary_search
function to perform a recursive binary search of the array. The function
should receive the search key, starting index and ending index as
arguments. If the search key is found, return its index in the array. If the
search key is not found, return –1.

In [ ]:
def binary_search(array, key, low, high):
  # Base case: not found
  if low > high:
    return -1

  mid = (low + high) // 2

  if array[mid] == key:
    return mid
  elif key < array[mid]:
    return binary_search(array, key, low, mid - 1)
  else:
    return binary_search(array, key, mid + 1, high)

numbers = [2, 5, 8, 12, 16, 23, 38, 56, 72]

print(binary_search(numbers, 23, 0, len(numbers) - 1))  # Found
print(binary_search(numbers, 10, 0, len(numbers) - 1))  # Not found

5
-1


11.19 (Quicksort) The recursive sorting technique called quicksort uses
the following basic algorithm for a one-dimensional array of values:
1. Partitioning Step: Take the first element of the unsorted array and
determine its final location in the sorted array (i.e., all values to
the left of the element in the array are less than the element, and
all values to the right of the element in the array are greater than
the element—we show how to do this below). We now have one
element in its proper location and two unsorted subarrays.
2. Recursive Step: Perform Step 1 on each unsorted subarray. Each
time Step 1 is performed on a subarray, another element is placed
in its final location of the sorted array, and two unsorted
subarrays are created. When a subarray consists of one element,
that element is in its final location (because a one-element array
is already sorted).
The basic algorithm seems simple enough, but how do we
determine the final position of the first element of each subarray?
As an example, consider the following set of values (the element
in bold is the partitioning element—it will be placed in its final
location in the sorted array):
37 2 6 4 89 8 10 12 68 45
Starting from the rightmost element of the array, compare each
element with 37 until an element less than 37 is found; then swap
37 and that element. The first element less than 37 is 12, so 37
and 12 are swapped. The new array is
12 2 6 4 89 8 10 37 68 45
Element 12 is in italics to indicate that it was just swapped with
37.
Starting from the left of the array, but beginning with the element
after 12, compare each element with 37 until an element greater
than 37 is found—then swap 37 and that element. The first
element greater than 37 is 89, so 37 and 89 are swapped. The
new array is
12 2 6 4 37 8 10 89 68 45
Starting from the right, but beginning with the element before 89,
compare each element with 37 until an element less than 37 is
found—then swap 37 and that element. The first element less
than 37 is 10, so 37 and 10 are swapped. The new array is
12 2 6 4 10 8 37 89 68 45
Starting from the left, but beginning with the element after 10,
compare each element with 37 until an element greater than 37 is
found—then swap 37 and that element. There are no more
elements greater than 37, so when we compare 37 with itself, we
know that 37 has been placed in its final location in the sorted
array. Every value to the left of 37 is smaller than it, and every
value to the right of 37 is larger than it.
Once the partition has been applied on the previous array, there
are two unsorted subarrays. The subarray with values less than 37
contains 12, 2, 6, 4, 10 and 8. The subarray with values greater
than 37 contains 89, 68 and 45. The sort continues recursively,
with both subarrays being partitioned in the same manner as the
original array.
Based on the preceding discussion, write recursive function
quick_sort_helper to sort a one-dimensional integer array.
The function should receive as arguments a starting index and an
ending index on the original array being sorted. Call this function
from a quick_sort function that receives just the original
array to sort.

11.20 (Determining Big O of Various Algorithms) Determine the Big O
for each of the following. You may need to research some of these items
online:
1. Get or set an item by index in a Python list.
2. Insert a new value in order in a Python sorted list.
3. Shell short an array.
4. Bubble sort an array.
5. Towers of Hanoi for n disks. [Hint: For n=1, 2, 3, 4, 5 or 6, the
number of operations is 1, 3, 7, 15, 31 or 63, respectively.] You’ll
see that the Big O for the Towers of Hanoi—O(2 )—is far worse
than that of the O(n ) sorts in this chapter. By the way, for 64
disks, the number of operations is 18,446,744,073,709,551,615.
If the priests could move one disk per second, it would take
584,942,417,355 years to move the stack of 64 disks!
6. Find all the permutations (unique arrangements) of n unique
items. [Hint: For the digits 1, 2 and 3 there are six permutations
—123, 132, 213, 231, 312 and 321.] For the digits 1, 2, 3 and 4
there are 24 permutations. For the digits 1, 2, 3, 4 and 5 there are
n
2
120 permutations. You’ll see that this Big O is far worse even
than that of the Towers of Hanoi.